hal/halx86/smp/smp.c

```diff
@@ -31,3 +61,126 @@ HalpSetupProcessorsTable(
     HalpProcessorIdentity[NTProcessorNumber].ProcessorPrcb = CurrentPrcb;
 }
+
+static
+VOID
+HalpMapAddressFlat(
+    _Inout_ PMMPDE PageDirectory,
+    _In_ PVOID VirtAddress,
+    _In_ PVOID TargetVirtAddress)
+{
+    if (TargetVirtAddress == NULL)
+        TargetVirtAddress = VirtAddress;
+
+    PMMPDE currentPde;
+
+    currentPde = &PageDirectory[MiAddressToPdeOffset(TargetVirtAddress)];
+
+    // Allocate a Page Table if there is no one for this address
+    if (currentPde->u.Long == 0)
+    {
+        PMMPTE pageTable = ExAllocatePoolZero(NonPagedPool, PAGE_SIZE, TAG_HAL);
+        ASSERT(pageTable);
+
+        currentPde->u.Hard.PageFrameNumber = MmGetPhysicalAddress(pageTable).QuadPart >> PAGE_SHIFT;
+        currentPde->u.Hard.Valid = TRUE;
+        currentPde->u.Hard.Write = TRUE;
+    }
+    
+    // Map the Page Table so we can add our VirtAddress there (hack around I/O memory mapper for that)
+    PHYSICAL_ADDRESS b = {.QuadPart = (ULONG_PTR)currentPde->u.Hard.PageFrameNumber << PAGE_SHIFT};
+
+    PMMPTE pageTable = MmMapIoSpace(b, PAGE_SIZE, MmCached);
+
+    PMMPTE currentPte = &pageTable[MiAddressToPteOffset(TargetVirtAddress)];
+    currentPte->u.Hard.PageFrameNumber = MmGetPhysicalAddress(VirtAddress).QuadPart >> PAGE_SHIFT;
+    currentPte->u.Hard.Valid = TRUE;
+    currentPte->u.Hard.Write = TRUE;
+
+    MmUnmapIoSpace(pageTable, PAGE_SIZE);
+
+    DPRINT("Map %p -> %p, PDE %u PTE %u\n",
+           TargetVirtAddress,
+           (PVOID)MmGetPhysicalAddress(VirtAddress).LowPart,
+           MiAddressToPdeOffset(TargetVirtAddress),
+           MiAddressToPteOffset(TargetVirtAddress));
+}
+
+static
+PHYSICAL_ADDRESS
+HalpInitTempPageTable(
+    _In_ PKPROCESSOR_STATE ProcessorState)
+{
+    PMMPDE pageDirectory = ExAllocatePoolZero(NonPagedPool, PAGE_SIZE, TAG_HAL);
+    ASSERT(pageDirectory);
+
+    // Map the low stub
+    HalpMapAddressFlat(pageDirectory, HalpLowStub, (PVOID)(ULONG_PTR)HalpLowStubPhysicalAddress.QuadPart);
+    HalpMapAddressFlat(pageDirectory, HalpLowStub, NULL);
+
+    // Map 32bit mode entry point
+    HalpMapAddressFlat(pageDirectory, &APEntry32, NULL);
+
+    // Map GDT
+    HalpMapAddressFlat(pageDirectory, (PVOID)ProcessorState->SpecialRegisters.Gdtr.Base, NULL);
+
+    // Map IDT
+    HalpMapAddressFlat(pageDirectory, (PVOID)ProcessorState->SpecialRegisters.Idtr.Base, NULL);
+
+    return MmGetPhysicalAddress(pageDirectory);
+}
+
+BOOLEAN
+NTAPI
+HalStartNextProcessor(
+    _In_ PLOADER_PARAMETER_BLOCK LoaderBlock,
+    _In_ PKPROCESSOR_STATE ProcessorState)
+{
+    if (HalpApicInfoTable.ProcessorCount > StartedProcessorCount)
+    { 
+        // Initalize the temporary page table
+        // TODO: clean it up after an AP boots successfully
+        ULONG_PTR initialCr3 = HalpInitTempPageTable(ProcessorState).QuadPart;
+
+        // Put the bootstrap code into low memory
+        RtlCopyMemory(HalpLowStub, &APEntry16,  ((ULONG_PTR)&APEntry16End - (ULONG_PTR)&APEntry16));
```

Comment:
There’s been a few attempts at making this work, currently it doesn’t for HAL but I have a few ideas if you don’t have a way you want to do this. Since we do similar for NTVDM

Reply:
> [...] Since we do similar for NTVDM

You mean, like this code? https://git.reactos.org/?p=reactos.git;a=blob;f=hal/halx86/generic/v86.S;hb=e627c3b00e7109b5110b94fb8253859a29f35d42
(and other that sets it up)

Reply:
Yeah Illen made some changes awhile back that inspired me a bit.
https://github.com/Ring-0-Productions/reactos/blob/Nightly-ARM32/subsystems/mvdm/asm16.cmake

Reply:
We might get away with "fake" 16 bit code, i.e. assemble as 64 bit , since most instructions should be identical.

Reply:
I would prefer we stay on the obvious safe side.

Reply:
Without linker support the alternative is to compile the code in 2 separate files, one for 16 bit, one for 64 bit, then using obj2bin on the 16 bit code, then converting it into c with bin2c, then compiling that into the code and use additional code to combine that with the 64 bit code and hope that all the offsets are correct. Doesn't sound any more "safe" to me.